### ISSUE TITLE
Align telemetry pipeline contract, storage, ML export, and admin reset endpoints

### ISSUE THREAD

**Issue by NeonButrfly** [NeonButrfly/tichuml#35]
## Summary
Align the telemetry pipeline end to end across self-play simulation, backend ingestion, database storage, ML export, evaluation summaries, and development/admin reset flows.

## Impact
Telemetry decisions are the canonical source for ML export and training, while events support debugging and replay analysis. If payload naming, validation, storage columns, or export assumptions diverge, training data becomes inconsistent and backend diagnostics become hard to query. Destructive development reset paths also need explicit safeguards so local iteration is possible without creating a production footgun.

## Reproduction Steps
1. Run self-play telemetry through the simulator and backend telemetry endpoints.
2. Inspect decision/event payload fields at emission, ingestion, database storage, and `ml/export_training_rows.py` export.
3. Compare provider naming, action/state metadata, schema/version fields, and extracted ML helper fields.
4. Attempt to clear telemetry or reset local app-owned database contents through a documented admin API.

## Expected vs Actual
Expected:
- one shared canonical telemetry contract for decisions and events
- deterministic required fields for decision/event ingestion
- malformed telemetry rejected early with structured errors
- decisions store raw JSON plus ML/query-friendly extracted fields
- events store raw payload plus useful extracted fields
- ML export reads canonical fields and reports rich metadata coverage
- development/admin clear/reset endpoints are explicit, guarded by env and confirmation token, and return row counts
- docs and prompt capture describe the contract and safeguards

Actual:
- telemetry is emitted and ingested, but the pipeline needs a single canonical contract, stronger validators, richer extracted storage fields, aligned export diagnostics, and explicit guarded destructive admin endpoints.

## Root Cause
Telemetry emission, ingestion, persistence, and export evolved as adjacent systems. The current shape works for baseline capture but lacks one enforceable contract and one service layer that preserves raw detail while exposing reliable scalar fields for ML and analysis.

## Proposed Fix
- add shared decision/event telemetry schemas, canonical naming, helpers, validation, and stable hashing
- update simulator emission and backend ingestion to use canonical payloads where practical
- extend database schema/migrations for raw JSONB and extracted scalar/indexed fields
- centralize telemetry insert, health/stat, clear, and reset operations in a service/repository layer
- update `ml/export_training_rows.py` for canonical fields and diagnostics
- add guarded `POST /api/admin/telemetry/clear`, `POST /api/admin/database/clear`, and `POST /api/admin/database/reset`
- update docs, prompt capture, and machine-readable alignment summary

## Validation Plan
- add/update telemetry validation tests for malformed decisions/events and chosen_action legality
- test admin clear/reset guard behavior and successful clears when safeguards are enabled
- run targeted backend telemetry/server tests
- run TypeScript build/checks covering shared, server, sim, and telemetry packages
- run or smoke-check `npm run ml:export`, `npm run ml:train`, and `npm run ml:evaluate` paths where feasible

## Prompt Capture Seed
Prompt request: fully line up the telemetry pipeline across sim runner, backend ingestion, database storage, ML export, and evaluation; add explicit guarded admin endpoints to clear telemetry/reset the development database; preserve rich decision/event detail for ML and analysis; do not regress sim/backend/export/train/evaluate workflows.

**Comment by NeonButrfly** [comment]
## Telemetry alignment implementation update

Implemented on branch `codex/telemetry-pipeline-alignment`.

### What changed
- Added canonical shared telemetry decision/event contract fields and validators.
- Added deterministic hashing and derived helper fields for decisions/events.
- Enforced decision consistency checks: actor-scoped legal actions, chosen action membership, `phase` vs `state_raw.phase`, and active actor consistency when available.
- Extended Postgres telemetry storage via migration `0003_telemetry_alignment.sql` with raw JSONB preservation, extracted scalar/query fields, indexes, and aggregate views.
- Updated backend repository insert/list paths to persist canonical provider context and ML/helper fields.
- Added guarded admin endpoints:
  - `POST /api/admin/telemetry/clear`
  - `POST /api/admin/database/clear`
  - `POST /api/admin/database/reset`
- Added env guard `ENABLE_DESTRUCTIVE_ADMIN_ENDPOINTS=true` plus required confirmation `CLEAR_TICHU_DB` via header or body.
- Updated sim runner, web telemetry emitter, backend providers, and LightGBM metadata capture for canonical provider/rich ML fields.
- Updated `ml/export_training_rows.py` to prefer canonical fields and expose `--diagnostics`.
- Added docs and prompt capture:
  - `docs/telemetry_contract.md`
  - `docs/admin_reset_endpoints.md`
  - `docs/telemetry_alignment_summary.json`
  - `docs/prompts/backend.md`

### Validation evidence
Passed:
- `npm run build`
- `npm run build:server`
- `npm run build:shared`
- `npm test -- --run tests/integration/backend-server.test.ts tests/integration/backend-telemetry.test.ts tests/integration/backend-client.test.ts tests/integration/server-heuristic-contract.test.ts` (21 tests)
- `npm test -- --run tests/integration/backend-server.test.ts tests/integration/backend-client.test.ts` (13 tests after final validator fallback polish)
- `npx eslint packages/shared/src/backend.ts apps/server/src/config/env.ts apps/server/src/providers/lightgbm-provider.ts apps/server/src/providers/provider-utils.ts apps/server/src/routes/router.ts apps/server/src/services/telemetry-repository.ts apps/server/src/utils/http.ts apps/sim-runner/src/self-play-batch.ts apps/web/src/backend/telemetry.ts tests/integration/backend-client.test.ts tests/integration/backend-server.test.ts tests/integration/backend-telemetry.test.ts tests/integration/server-heuristic-contract.test.ts`
- `git diff --check`
- `npm run ml:export -- --help`
- `npm run ml:train -- --help`

Caveats:
- Root `npm run lint` still scans `.venv/Lib/site-packages/sklearn/...` and fails on third-party JS outside this task.
- Full `npm run sim -- --games 1 --provider local --telemetry false --quiet` and `npm run ml:evaluate -- --games 1 --provider local --telemetry false --quiet --output eval/results/telemetry-alignment-smoke.json` exceeded a 124 second validation timeout. Follow-up tracked as #36.

### Status
Implementation is complete for #35, but I am leaving the issue open pending live DB migration/application validation and the sim/evaluate runtime follow-up in #36.

**Comment by NeonButrfly** [comment]
## Telemetry alignment follow-up validation

Additional hardening completed on branch `codex/telemetry-pipeline-alignment` after the snapshot commit.

### Added after audit
- Persisted and exported `chosen_action_is_legal` as a first-class decision scalar.
- Added `explanation_quality_level` (`none`, `basic`, `scored`, `featured`) for coarse rich-signal coverage.
- Expanded telemetry health stats with unique/duplicate state and legal-action hash counts, provider/phase/seat/type aggregates, legal chosen-action counts, wish/pass counts, and freshness timestamps.
- Added duplicate-state and duplicate-legal-action aggregate views.
- Added richer decision metadata where available: seed, latency, current lead/current combination summary, wish active/satisfiable status, and no-legal-wish-fulfillment marker.
- Updated exporter diagnostics to report legal chosen-action coverage and malformed legacy decisions filtered from training rows.
- Added validation coverage for phase/actor contradictions and deterministic canonical JSON hashing.

### Validation evidence
Passed:
- `npm run build:shared`
- `npm run build:server`
- `npm run build:sim-runner`
- `npm run build`
- `npm test -- --run tests/integration/backend-server.test.ts tests/integration/backend-telemetry.test.ts tests/integration/backend-client.test.ts tests/integration/server-heuristic-contract.test.ts` (23 tests)
- `npx eslint packages/shared/src/backend.ts apps/server/src/services/decision-service.ts apps/server/src/services/telemetry-repository.ts apps/server/src/providers/provider-utils.ts apps/sim-runner/src/self-play-batch.ts tests/integration/backend-server.test.ts tests/integration/server-heuristic-contract.test.ts`
- `git diff --check`
- `python -m json.tool docs/telemetry_alignment_summary.json`
- `python -m py_compile ml/export_training_rows.py`
- `npm run ml:export -- --help`
- `npm run ml:train -- --help`
- `npm run sim -- --games 0 --provider local --telemetry false --quiet`
- `npm run ml:evaluate -- --games 0 --provider local --telemetry false --quiet --output eval/results/telemetry-alignment-smoke.json`

Caveats:
- `npm run db:migrate:status` could not connect to Postgres (`ECONNREFUSED` on localhost:5432), and Docker Desktop is not running on this machine, so live DB migration execution remains unverified here.
- `npm run sim -- --help` and `npm run ml:evaluate -- --help` are not valid help paths today; those CLIs attempted to run and failed on backend reachability.

### Status
Leaving #35 open pending live Postgres migration/application validation. Code, docs, tests, and exporter alignment are complete within the available local environment.

**Comment by NeonButrfly** [comment]
## Producer-side telemetry centralization update

Implemented the focused refactor requested in this pass:

- `packages/telemetry` is now the authoritative producer-side subsystem for decision/event payload builders, source adapters, telemetry config normalization, minimal/full/adaptive mode selection, byte measurement, downgrade/skip policy, shared POST behavior, failure classification, strict-mode errors, and structured diagnostics.
- Normal gameplay now routes through `apps/web/src/backend/telemetry.ts -> @tichuml/telemetry -> /api/telemetry/*`.
- Simulator/selfplay/controller telemetry now routes through `apps/sim-runner/src/self-play-batch.ts -> @tichuml/telemetry -> /api/telemetry/*`.
- Legacy web client telemetry helpers now delegate to `@tichuml/telemetry` instead of owning raw telemetry POST behavior.
- Source tags are applied via `metadata.source` / `metadata.telemetry_source` for `gameplay`, `selfplay`, and `controller`.
- Non-strict telemetry failures return structured results and are counted/logged without throwing into gameplay/selfplay/controller paths; strict mode surfaces `TelemetryError` for targeted debugging.
- Full payloads can still carry rich training fields (`state_raw`, `state_norm`, `explanation`, `candidateScores`, `stateFeatures`); oversize full payloads downgrade to minimal, then skip locally if still too large.
- Docs and prompt capture updated to point future producers at `packages/telemetry`.

Validation evidence:

- `npm run build` passed.
- `npx vitest run tests/integration/central-telemetry.test.ts tests/integration/backend-telemetry.test.ts tests/integration/backend-client.test.ts tests/integration/server-heuristic-contract.test.ts -t "central telemetry subsystem|backend telemetry decision tracking|backend client error classification|telemetry"` passed: 4 files, 17 tests passed, 13 skipped by filter.
- `npx eslint packages/telemetry/src apps/sim-runner/src/self-play-batch.ts apps/web/src/backend/telemetry.ts apps/web/src/backend/client.ts tests/integration/central-telemetry.test.ts tests/integration/server-heuristic-contract.test.ts` passed.
- `git diff --check` passed.
- Full `npm test` was attempted but timed out at 184s in the local Codex environment, so I am not claiming full-suite completion.

Audit:

- Previous producers: simulator/selfplay in `apps/sim-runner/src/self-play-batch.ts`, normal gameplay in `apps/web/src/backend/telemetry.ts`, web backend-client telemetry helpers in `apps/web/src/backend/client.ts`, backend ingestion routes in `apps/server/src/routes/router.ts`.
- New path: all producer-side builders/posting for sim, controller, gameplay, and compatibility helpers go through `@tichuml/telemetry`.
- Remaining direct `/api/telemetry/*` fetches found by audit are backend route/ingestion tests, not production producers.
- Normal gameplay and selfplay now feed the same backend decision/event tables with shared payload policy and source metadata.

**Comment by NeonButrfly** [comment]
Final local commit for this producer-side centralization pass:

- pre-change snapshot commit: `f1426e2b7243c6a9d8dac2c90f4fd4d929840d00` (`chore: snapshot before telemetry/controller changes`)
- implementation commit: `e2cc0eb7383db3c6ec4076f914d9d3b70736dc63` (`refactor: centralize telemetry pipeline`)

The earlier validation/audit comment in this issue records the command evidence and the full `npm test` timeout caveat.

**Comment by NeonButrfly** [comment]
Implementation update from commit 58013d088494be8640cc9ac6fd7cace163a1114d (`fix: harden runtime telemetry controller flow`).

Summary:
- Kept `@tichuml/telemetry` as the authoritative producer-side path and removed the remaining backend decision manual telemetry payload shaping from `apps/server/src/providers/provider-utils.ts` by delegating to the shared builder.
- Added central client timeout/retry config (`TELEMETRY_POST_TIMEOUT_MS`, retry defaults) and preserved non-strict best-effort behavior.
- Changed telemetry ingest routes to validate synchronously, enqueue accepted payloads, return 202, and expose queue stats through `/api/telemetry/health`.
- Added bounded ingest queue diagnostics for queue pressure and persistence failures.

Validation evidence:
- `npm run build:shared` passed.
- `npm run build:telemetry` passed.
- `npm run build:server` passed.
- `npm run build:sim-runner` passed.
- `npm run build:web` passed.
- `npx vitest run tests/integration/telemetry-ingest-queue.test.ts tests/integration/central-telemetry.test.ts tests/integration/backend-server.test.ts tests/integration/runtime-admin-service.test.ts tests/integration/server-config.test.ts tests/integration/server-heuristic-contract.test.ts` passed: 6 files, 72 tests.
- `git diff --check` and `git diff --cached --check` passed.

Remaining reality/caveat:
- Full `npm test` and local `npm run sim -- --games 1 --provider local --telemetry false --quiet --progress false` still exceeded local time bounds in this workspace, so live/full-run closure should stay open until host-side validation confirms long-run simulator behavior and DB persistence under real load.
